openvdb/points/PointMask.h

```diff
@@ -0,0 +1,384 @@
+                    const std::vector<Name>& includeGroups = std::vector<Name>(),
+                    const std::vector<Name>& excludeGroups = std::vector<Name>());
+
+
+////////////////////////////////////////
+
+
+namespace point_mask_internal {
+
+/// @brief Combines multiple grids into one by stealing leaf nodes and summing voxel values
+/// This class is designed to work with thread local storage containers such as tbb::combinable
+template<typename GridT>
+struct GridCombinerOp
+{
+    using CombinableT = typename tbb::combinable<GridT>;
+
+    using TreeT = typename GridT::TreeType;
+    using LeafT = typename TreeT::LeafNodeType;
+    using ValueType = typename TreeT::ValueType;
+    using SumOp = tools::valxform::SumOp<typename TreeT::ValueType>;
+
+    GridCombinerOp(GridT& grid)
+        : mTree(grid.tree()) {}
+
+    void operator()(const GridT& grid)
+    {
+        for (auto leaf = grid.tree().beginLeaf(); leaf; ++leaf) {
+            auto* newLeaf = mTree.probeLeaf(leaf->origin());
+            if (!newLeaf) {
+                // if the leaf doesn't yet exist in the new tree, steal it
+                auto& tree = const_cast<GridT&>(grid).tree();
+                mTree.addLeaf(tree.template stealNode<LeafT>(leaf->origin(),
+                    zeroVal<ValueType>(), false));
+            }
+            else {
+                // otherwise increment existing values
+                for (auto iter = leaf->cbeginValueOn(); iter; ++iter) {
+                    newLeaf->modifyValue(iter.getCoord(), SumOp(*iter));
+                }
+            }
+        }
+    }
+
+private:
+    TreeT& mTree;
+}; // struct GridCombinerOp
+
+
+/// @brief Compute scalar grid from PointDataGrid while evaluating the point filter
+template <typename GridT, typename PointDataGridT, typename FilterT>
+struct PointsToScalarOp
+{
+    using TreeT = typename GridT::TreeType;
+    using LeafManagerT = typename tree::LeafManager<TreeT>;
+    using ValueT = typename TreeT::LeafNodeType::ValueType;
+
+    PointsToScalarOp(const PointDataGridT& grid,
+                  const FilterT& filter)
+        : mPointDataAccessor(grid.getConstAccessor())
+        , mFilter(filter) { }
+
+    void operator()(const typename LeafManagerT::LeafRange& range) const {
+
+        for (auto leaf = range.begin(); leaf; ++leaf) {
+
+            const auto* const pointLeaf =
+                mPointDataAccessor.probeConstLeaf(leaf->origin());
+
+            // assumes matching topology
+            assert(pointLeaf);
+
+            for (auto value = leaf->beginValueOn(); value; ++value) {
+                const Index64 count = points::iterCount(
+                    pointLeaf->beginIndexVoxel(value.getCoord(), mFilter));
+                if (count > Index64(0)) {
+                    value.setValue(static_cast<ValueT>(count));
+                }
+                else {
+                    // disable any empty voxels
+                    value.setValueOn(false);
+                }
```

Comment:
Isn't it assumed that active voxels will always have points? Could this not be left down to the user to deactivate 0 voxels?

```
for (auto value = leaf->beginValueOn(); value; ++value) {
    const Index64 count = points::iterCount(
        pointLeaf->beginIndexVoxel(value.getCoord(), mFilter));
    value.setValue(static_cast<ValueT>(count));
}

```

and users would call:

```
tools::deactivate(countTree, 0);

```

As it's creating a "mask" it should at least explicitly define the workflow for 0 voxels as at the moment an int count tree will deactivate them but a bool mask tree won't

Reply:
The reason I didn't do that is because it needs an extra iteration and comparison which seems wasteful.

I don't think you're right about a bool mask tree not deactivating the voxels though. This is a little confusing I find, but this logic is calling the setValueOn() method on the Iterator, which evaluates this:

```
void setValueOn(bool on = true) const { parent().setValueMask(this->pos(), on); }
```

which in turn calls the following method on LeafNodeBool:

```
void setValueMask(Index n, bool on) { mValueMask.set(n, on); }
```

I believe this technique does correctly deactivate voxels on bool trees, int trees and mask trees.

Reply:
Yes sorry I wasn't clear, there is an early exit at line 251 for bool trees which means the topology could differ

Reply:
Yep sorry I wasn't clear, there's an early exit at line 251 which means bool topology would differ

Reply:
Okay let me put this another way - currently PointsToScalarOp checks to see if the voxel contains points even if the voxel is active. This means that the op assumes that active voxels could contain 0 points. So the result of this, if given a pdg topology with active empty voxels is a different, slightly modified topology. Given a pdg called "grid" with active zero voxels with a group "group1" which all points belong to, calling convertPointsToScalar(grid) vs convertPointsToScalar(grid, "group1") will give different results, as the prior will early exist before running the op

Reply:
If we assume the topology to be "correct" at one place we should assume it in the other and not check for 0 point counts for consistency

Reply:
I see what you're getting at, though there is one thing I think you're missing. In the case where there are valid points, the call to setValue() does not activate voxels, it *only* sets the value. Hence if you were to set the active state of all the voxels prior to this method to inactive, you would still end up with the same mask entirely inactive topology.

I think if you have active zero voxels, you should de-activate them prior to calling this method as all methods should be allowed to expect *correct* point data grids for performance reasons. If you had a different interpretation of what an active voxel was and didn't want them de-activated, you could always extract the non-filtered topology and union it with the resulting topology. As I say, I see this being a rare case, so I'm not convinced the API should account for it provided you can easily work around it if you were doing something unusual.

Reply:
Yep I understand the implementation, the second paragraph of your reply is somewhat what I was originally getting at. Still not sure I've completely made my point clear - I'm all for the public API to operator under the assumption that the point data grids being used have "expected" values and active topology. My point is that this different in branching between using groups and not using groups for the construction of a bool value type grid is not consistent. On the one had we early exit without checking values in relation to active values, where as on the other we explicitly check 

```
if (count > Index64(0)) 
```
on line 156 and deactivate them otherwise. Note that that operation only loops over active voxels, so if we are assuming correct topology why do we need to check for zero values? If we are leaving the early exit in, then a better implementation would be:
```
           for (auto value = leaf->beginValueOn(); value; ++value) {
                const Index64 count = points::iterCount(
                    pointLeaf->beginIndexVoxel(value.getCoord(), mFilter));
                value.setValue(static_cast<ValueT>(count));
           }
```

This is now consistent in the assumption of the topology in relation to the activity. As you say, users can correct the grids prior to running this, or alternatively deactivate them in post as I suggested - but this is completely down to the caller.

Reply:
Further we shouldn't need to actually count points for boolean or mask grids - could this be templated with simply:

```
    if (std::is_same<bool, T>::value) value.setValue(true);
```

Reply:
Fair enough, in order to keep the API consistent, I'm happy to leave it up to the user to do the deactivation and pruning. I've introduced what you've suggested for point counts and explicitly added deactivation and pruning to the OpenVDB Points Convert SOP.

We can't avoid counting, because we still need to evaluate the filter (if it's in use) in order to know whether to set the value to true or false. The value still depends on whether there are valid points or not in each voxel.

Reply:
Perhaps we could branch on a NullFilter? This way you could avoid counting, match the bool topology early exit AND provide a deactivated topology on the fly as you're currently doing as the bool early exit only exits if usegroups is false

Reply:
NullFilter is always used if usegroups is false, so the early exit should catch this? In which cases do you not want to count?

Reply:
Ah sorry, I meant branching for non mask types such that zero values remain active in count grids (for NullFilter don't check count > 0), as this follows the assumption of active voxels not being empty. Count grids would then match mask grids created with the same point topology. Perhaps I'm labouring this too heavily though.

Reply:
Hmm, this doesn't seem to be working as expected. The value setting methods on the iterator are modifying the active state directly and not the bool value meaning the bool values are not set when copying the topology, which rather complicates things.

I'm going to submit this portion of the routine as it is in it's working state and we'll have to come back and look at this method to see if what you're asking is feasible.